File: Plugins/QuorumPlugins/src/main/java/plugins/quorum/Libraries/System/DateTime.java
package plugins.quorum.Libraries.System;

import java.time.*;
import java.time.format.DateTimeFormatter;
import java.util.Calendar;
import java.util.Date;
import java.util.SimpleTimeZone;
import java.util.TimeZone;

/**
 * The default implementation of the date time plugin for the compiler.
 * 
 * @author Andreas Stefik
 */
public class DateTime {
    public java.lang.Object me_ = null;

    private boolean now = true; // if false, user has called SetEpochTime()
    private TimeZone timezone = new SimpleTimeZone(Calendar.getInstance().getTimeZone().getOffset(System.currentTimeMillis()), "local timezone");
    private long epochTime = 0; // only used if user has called SetEpochTime()*/
    private Calendar cal;

    ZoneId zoneId = ZoneId.systemDefault();
    public ZonedDateTime zDateTime = ZonedDateTime.now(ZoneId.systemDefault());

    public int GetYear() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getYear();
    }

    public int GetMonth() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getMonthValue();
    }

    public int GetDayOfMonth() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getDayOfMonth();
    }

    public int GetDayOfWeek() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getDayOfWeek().getValue();
    }

    public int GetHour() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getHour();
    }

    public int GetMinute() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getMinute();
    }

    public int GetSecond() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.getSecond();
    }

    public int GetTimeZone() {
        ZonedDateTime temp = ZonedDateTime.now(zoneId);
        ZoneOffset offset = temp.getOffset();
        return (offset.getTotalSeconds()) / 3600;
    }

    public boolean IsDaylightSavings() {
        return zoneId.getRules().isDaylightSavings(zDateTime.toInstant());
    }

    public double GetEpochTime() {
        if (now)
            zDateTime = zDateTime.now(zoneId);
        return zDateTime.toInstant().toEpochMilli();
    }

    public void SetEpochTime(double epochTime) {
        // Set the internal representation of time to the given epoch time.
        now = false;
        Instant instant = Instant.ofEpochMilli((long)epochTime);
        zDateTime = ZonedDateTime.ofInstant(instant, zoneId);
    }

    public void SetTimeZone(int timeZoneOffset) {
        timezone = new SimpleTimeZone(timeZoneOffset * 1000 * 60 *60, "custom time zone");
        zoneId = timezone.toZoneId();
    }

    public double Now() {
        return System.currentTimeMillis();
    }

    //output functions
    public String GetTimestamp() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME);
    }

    public String ToText() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.format(DateTimeFormatter.RFC_1123_DATE_TIME);
    }

    public String TimeToText() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.format(DateTimeFormatter.ISO_LOCAL_TIME);
    }

    public String DateToText() {
        if (now)
            zDateTime = zDateTime.now(zoneId);

        return zDateTime.format(DateTimeFormatter.ISO_LOCAL_DATE);
    }

    public String GetMonthName() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMMM");
        return zDateTime.format(formatter);
    }

    public String GetShortMonthName() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM");
        return zDateTime.format(formatter);
    }

    public String GetDayOfWeekName() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("EEEE");
        return zDateTime.format(formatter);
    }

    public String GetShortDayOfWeekName() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("EEE");
        return zDateTime.format(formatter);
    }
    // parse functions
    public void ParseTimestamp(String timestamp) {
        now = false;
        timestamp = timestamp.replace(' ', 'T');
        LocalDateTime temp = LocalDateTime.parse(timestamp, DateTimeFormatter.ISO_LOCAL_DATE_TIME);
        zDateTime = temp.atZone(zoneId);
    }

    public void ParseDate(String date) {
        now = false;
        LocalDate temp = LocalDate.parse(date, DateTimeFormatter.ISO_LOCAL_DATE);
        zDateTime = temp.atStartOfDay().atZone(zoneId);
    }

    public void ParseAmericanDate(String date) {
        now = false;
        LocalDate temp = LocalDate.parse(date, DateTimeFormatter.ofPattern("M/d/[uuuu][uu]"));
        zDateTime = temp.atStartOfDay().atZone(zoneId);
    }

    public void ParseTime(String time, boolean keepDate) {
        LocalTime temp = LocalTime.parse(time, DateTimeFormatter.ISO_LOCAL_TIME);
        LocalDate date;
        if (keepDate) {
            date = zDateTime.toLocalDate();
        } else {
            date = LocalDate.now();
        }
        zDateTime = ZonedDateTime.of(date, temp, zoneId);
    }

    // compare functions
    public boolean IsBefore(quorum.Libraries.System.DateTime_ dateTime) {
        quorum.Libraries.System.DateTime temp = (quorum.Libraries.System.DateTime) dateTime;
        DateTime dt = temp.plugin_;
        ZonedDateTime compared = dt.zDateTime;
        if (now && dt.now)
            return false;
        if (now)
            zDateTime = ZonedDateTime.now(zoneId);
        if (dt.now)
            compared = ZonedDateTime.now(zoneId);
        return zDateTime.isBefore(dt.zDateTime);
    }

    public boolean IsAfter(quorum.Libraries.System.DateTime_ dateTime) {
        quorum.Libraries.System.DateTime temp = (quorum.Libraries.System.DateTime) dateTime;
        DateTime dt = temp.plugin_;
        ZonedDateTime compared = dt.zDateTime;
        if (now && dt.now)
            return false;
        if (now)
            zDateTime = ZonedDateTime.now(zoneId);
        if (dt.now)
            compared = ZonedDateTime.now(zoneId);
        return zDateTime.isAfter(dt.zDateTime);
    }

    // math functions
    public void AddYears(int years) {
        now = false;
        zDateTime = zDateTime.plusYears(years);
    }

    public void AddMonths(int months) {
        now = false;
        zDateTime = zDateTime.plusMonths(months);
    }

    public void AddWeeks(int weeks) {
        now = false;
        zDateTime = zDateTime.plusWeeks(weeks);
    }

    public void AddDays(int days) {
        now = false;
        zDateTime = zDateTime.plusDays(days);
    }

    public void AddHours(int hours) {
        now = false;
        zDateTime = zDateTime.plusHours(hours);
    }

    public void AddMinutes(int minutes) {
        now = false;
        zDateTime = zDateTime.plusMinutes(minutes);
    }

    public void AddSeconds(int seconds) {
        now = false;
        zDateTime = zDateTime.plusSeconds(seconds);
    }

    public void SubtractYears(int years) {
        now = false;
        zDateTime = zDateTime.minusYears(years);
    }

    public void SubtractMonths(int months) {
        now = false;
        zDateTime = zDateTime.minusMonths(months);
    }

    public void SubtractWeeks(int weeks) {
        now = false;
        zDateTime = zDateTime.minusWeeks(weeks);
    }

    public void SubtractDays(int days) {
        now = false;
        zDateTime = zDateTime.minusDays(days);
    }

    public void SubtractHours(int hours) {
        now = false;
        zDateTime = zDateTime.minusHours(hours);
    }

    public void SubtractMinutes(int minutes) {
        now = false;
        zDateTime = zDateTime.minusMinutes(minutes);
    }

    public void SubtractSeconds(int seconds) {
        now = false;
        zDateTime = zDateTime.minusSeconds(seconds);
    }
}

File: Quorum/Library/Standard/Plugins/javascript/Libraries/System/DateTime.js
function plugins_quorum_Libraries_System_DateTime_() {
    
    this.date = new Date();
    this.now = true

    this.GetYear = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getFullYear();
    };  

    this.GetMonth = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getMonth() + 1;
    };  
    
    this.GetDayOfMonth = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getDate();
    };  
    
    this.GetDayOfWeek = function () {
        // this is keep it to the ISO 8601 standard
        // Days go from 1-7 where 1 is Monday and 7 is Sunday
        if(this.now) {
            this.date = new Date();
        }
        let day = date.getDay();
        if(day == 0) {
            day = 7;
        }
        return day;
    };  
    
    this.GetHour = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getHour();
    };  
    
    this.GetMinute = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getMinute();
    };  
    
    this.GetSecond = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getSecond();
    };  
    
    this.GetTimeZone = function () {
        if(this.now) {
            this.date = new Date();
        }
        return date.getTimezoneOffset();
    };  
    
    this.IsDaylightSavings = function () {
        if(this.now) {
            this.date = new Date();
        }
        // Jan and July are never in the same timezone if DST is observed
        let jan = new Date(d.getFullYear(), 0, 1).getTimezoneOffset();
        let jul = new Date(d.getFullYear(), 6, 1).getTimezoneOffset();
        return Math.max(jan, jul) !== this.date.getTimezoneOffset();   
    };  
    
    this.GetEpochTime = function () {
        if(this.now) {
            this.date = new Date();
        }
        return this.date.getTime();
    };  
    
    this.SetEpochTime$quorum_number = function (value) {
        this.now = false;
        this.date.setTime(value);
    };  
    
    this.SetTimeZone$quorum_integer = function (value) {
        //not supported in JS
    };
    
    this.Now = function () {
        return Date.now();
    };
    
    this.GetTimestamp = function () {
        return this.date.toISOString();
    };

    this.ToText = function () {
        return this.date.toString();
    };

    this.DateToText = function () {
        return this.date.toLocaleDateString();
    };

    this.TimeToText = function () {
        return this.date.toTimeString();
    };

    this.GetMonthName = function () {
        return this.date.toLocaleString('default', { month : 'long' });
    };

    this.GetShortMonthName = function () {
        return this.date.toLocaleString('default', { month : 'short' });
    };

    this.GetDayOfWeekName = function () {
        return this.date.toLocaleString('default', { weekday : 'long'});
    }

    this.GetShortDayOfWeekName = function () {
        return this.date.toLocaleString('default', { weekday : 'short'});
    }
    // Parsing
    // parse is implementation specific so it might not work
    // as expected or even on different browsers 
    this.ParseTimestamp$quorum_text = function (value) {
        this.now = false;
        value = value.trim();
        value = value.replace(' ', 'T');
        this.date = new Date(Date.parse(value));
        if(isNaN(this.date))
        {
            let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a timestamp.");
            throw exceptionInstance_;
        }
       	// check further
        let originalDate = value.substring(0,10);
        let arr = originalDate.split('-');
        if(arr.length != 3)
        {
        	let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a timestamp.");
            throw exceptionInstance_;
        }
        let originalTime = value.substring(11,20);
        arr = originalTime.split(':');
        if(arr.length != 3)
        {
        	let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a timestamp.");
            throw exceptionInstance_;
        }
    };

    this.ParseDate$quorum_text = function (value) {
        this.now = false;
        value = value.trim();
        this.date = new Date(Date.parse(value));

        if(isNaN(this.date))
        {
            let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a date.");
            throw exceptionInstance_;
        }
        // check further
        let arr = value.split('-');
        if(arr.length != 3 && value.length != 10)
        {
        	let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a date.");
            throw exceptionInstance_;
        }
    };

    this.ParseAmericanDate$quorum_text = function (value) {
        this.now = false;
        value = value.trim();
        this.date = new Date(Date.parse(value));

        if(isNaN(this.date))
        {
            let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a date.");
            throw exceptionInstance_;
        }
        // check further
        let arr = value.split('/');
        if(arr.length != 3 && value.length > 10)
        {
            let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Could not parse '"+value+"' as a date.");
            throw exceptionInstance_;
        }
    };

    this.ParseTime$quorum_text = function (value, keepDate) {
        this.now = false;
        // Not supported
        let exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
        exceptionInstance_.SetErrorMessage$quorum_text("Quorum running in Javascript does not support parsing a time alone.");
        throw exceptionInstance_;
    };

    this.IsBefore$quorum_Libraries_System_DateTime_ = function (value) {
        return date < value.plugin_.date;
    };

    this.IsAfter$quorum_Libraries_System_DateTime_ = function () {
        return date > value.plugin_.date;
    };

    // Math functions

    // JS Date doesn't have math functions but the Set functions account
    // for overflow and negative numbers so they are used instead

    this.AddYears$quorum_integer = function(value) {
        this.now = false;
        this.date.setFullYear(date.getFullYear() + value);
    };

    this.AddMonths$quorum_integer = function(value) {
        this.now = false;
        this.date.setMonth(this.date.getMonth() + value);
    };

    this.AddWeeks$quorum_integer = function(value) {
        this.now = false;
        this.date.setDay(this.date.getDay() + (value * 7));
    };

    this.AddDays$quorum_integer = function(value) {
        this.now = false;
        this.date.setDay(this.date.getDay() + value);
    };

    this.AddHours$quorum_integer = function(value) {
        this.now = false;
        this.date.setHours(this.date.getHours() + value);
    };

    this.AddMinutes$quorum_integer = function(value) {
        this.now = false;
        this.date.setMinutes(this.date.getMinutes() + value);
    };

    this.AddSeconds$quorum_integer = function(value) {
        this.now = false;
        this.date.setSeconds(this.date.getSeconds() + value, 0);
    };

    this.SubtractYears$quorum_integer = function(value) {
        this.now = false;
        this.date.setFullYear(date.getFullYear() - value);
    };

    this.SubtractMonths$quorum_integer = function(value) {
        this.now = false;
        this.date.setMonth(this.date.getMonth() - value);
    };

    this.SubtractWeeks$quorum_integer = function(value) {
        this.now = false;
        this.date.setDay(this.date.getDay() - (value * 7));
    };

    this.SubtractDays$quorum_integer = function(value) {
        this.now = false;
        this.date.setDay(this.date.getDay() - value);
    };

    this.SubtractHours$quorum_integer = function(value) {
        this.now = false;
        this.date.setHours(this.date.getHours() - value);
    };

    this.SubtractMinutes$quorum_integer = function(value) {
        this.now = false;
        this.date.setMinutes(this.date.getMinutes() - value);
    };

    this.SubtractSeconds$quorum_integer = function(value) {
        this.now = false;
        this.date.setSeconds(this.date.getSeconds() - value, 0);
    };

}

